src/mlpack/methods/ann/layer/minibatch_discrimination.hpp

```diff
@@ -0,0 +1,190 @@
+/**
+ * @file minibatch_discrimination.hpp
+ * @author Saksham Bansal
+ *
+ * Definition of the MiniBatchDiscrimination layer class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MINIBATCH_DISCRIMINATION_HPP
+#define MLPACK_METHODS_ANN_LAYER_MINIBATCH_DISCRIMINATION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "layer_types.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the MiniBatchDiscrimination layer. MinibatchDiscrimination
+ * is a layer of the discriminator that allows the discriminator to look at
+ * multiple data examples in combination and perform what is called as
+ * mini-batch discrimination.
+ * This helps prevent the collapse of the generator parameters to a setting
+ * where it emits the same point. This happens because normally a
+ * discriminator will process each example independently and there will be
+ * no mechanism to diversify the outputs of the generator.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Goodfellow2016,
+ *   author  = {Tim Salimans, Ian Goodfellow, Wojciech Zaremba, Vicki Cheung,
+ *              Alec Radford, Xi Chen},
+ *   title   = {Improved Techniques for Training GANs},
+ *   year    = {2016},
+ *   url     = {https://arxiv.org/abs/1606.03498},
+ * }
+ * @endcode
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class MiniBatchDiscrimination
+{
+ public:
+  //! Create the MiniBatchDiscrimination object.
+  MiniBatchDiscrimination();
+
+  /**
+   * Create the MiniBatchDiscrimination layer object using the specified
+   * number of units.
+   *
+   * @param inSize The number of input units.
+   * @param outSize The number of output units.
+   * @param features The number of features to compute for each dimension.
+   */
+  MiniBatchDiscrimination(const size_t inSize,
```

Comment:
We could remove the constructor above if we set a default value for each parameter here.

Reply:
Yes, I see.I don't think it is possible to assign a default value to `inSize` though.

Reply:
Right, I think for the other default constructor you used 0, which I think isn't a good default value as well. So maybe it makes sense to remove the default constructor?

Reply:
I think the default constructor is required for serialisation. Let me know if that is wrong. 
I also see many layers that have default constructors with no parameters initialised.